lib/tcpip/tcpip_error_handler.h

```diff
@@ -0,0 +1,367 @@
+ * destruction. The list of synchronization primitives resetted here was
+ * extracted through a manual study of the compartment's code-base: this may
+ * therefore break if new releases of FreeRTOS+TCP introduce new locks. In the
+ * future, we may want to come up with a more systematic approach.
+ *
+ * This function is designed to be robust against most types of compartment
+ * corruption, however we do assume that:
+ * - 'reset-critical' data has not been corrupted
+ * - the control-flow of threads in the compartment has not been altered
+ */
+extern "C" void reset_network_stack_state()
+{
+	const bool isUserThread = thread_id_get() != networkThreadID;
+	const bool isIpThread   = !isUserThread;
+
+	if (isUserThread)
+	{
+		DebugErrorHandler::log(
+		  "User thread TCP/IP stack error handler called!");
+		userThreadCount--;
+	}
+	else
+	{
+		DebugErrorHandler::log(
+		  "Network thread TCP/IP stack error handler called!");
+	}
+
+	// Manually unlock the sealed sockets lock list if it was held.
+	if (sealedSocketsListLock.get_owner_thread_id() == thread_id_get())
+	{
+		DebugErrorHandler::log("The sealed sockets lock was held by the "
+		                       "crashing thread. Forcefully unlocking it.");
+		sealedSocketsListLock.unlock();
+	}
+
+	// Set the currently restarting flag. This will do several things:
+	// 1. ensure that only one call to this error handler triggers a reset
+	// 2. ensure that no thread enters the compartment while we are
+	//    restarting
+	// 3. reset the network thread whenever it wakes up
+	uint8_t expected = NotRestarting;
+	if (!restartState.compare_exchange_strong(expected, Restarting))
+	{
+		// `expected` now contains a snapshot of `restartState`.
+		if (isIpThread && ((expected & IpThreadKicked) != 0))
+		{
+			// Currently recovering from a crash that happens
+			// during the reset process is not possible. It is not
+			// clear if we ever really want to do that: we will
+			// only crash during reset if 1) there is a bug in the
+			// reset code, or 2) there is some global data that we
+			// cannot reset and which is corrupted. In either case,
+			// re-reseting the same way will not make the situation
+			// better.
+			DebugErrorHandler::log("The network thread crashed while "
+			                       "restarting. This may be unrecoverable.");
+		}
+
+		// Another instance of the error handler is running, do not do
+		// anything.
+		return;
+	}
+
+	DebugErrorHandler::log("Reset-ing the network stack.");
+
+	// We need to acquire the sealed sockets lock because we do not want
+	// the sealed sockets list to be in an inconsistent state when we go
+	// over it.
+	//
+	// Waiting to acquire the lock is fine, as we know that any thread
+	// which holds the it will eventually release it, either 1) exiting the
+	// critical section, or 2) crashing into it, in which case we (the
+	// error handler) will manually unlock it (see manual unlock above).
```

Comment:
In theory, a crashing thread may corrupt the state of this lock, but that would require some memory-safety bug which, hopefully, can't happen.

Reply:
Correct, I do not know how attackers could corrupt the lock state without violating spatial and temporal memory safety. And if attackers manage that, it's something we need to fix in the locks I guess!